### PR TITLE
Add rich text description to woo product sync to meta

### PR DESCRIPTION
**Pull request by devbodaghe**
Changes proposed in this Pull Request:

This PR introduces a new feature to the Facebook WooCommerce Plugin, allowing users to add rich text descriptions to their products. With this update, we can seamlessly synchronize the rich text description field with the Facebook Commerce Manager platform.

Screenshot 2024-11-04 at 11 21 23
Screenshot 2024-11-04 at 11 22 26
Screenshot 2024-11-04 at 11 22 41

### CODE REVIEW

**Comment by mshymon** [comment]
You use "rtd" in the title, I am not sure how many people would grasp what is this change is about by reading the title, I didn't at first.

File: includes/fbproduct.php [hunk 1]
```diff
@@ -392,6 +411,7 @@ public function set_use_parent_image( $setting ) {
 
 	public function get_fb_description() {
 		$description = '';
+		$this->fb_description;
```
Comment (mshymon) [inline includes/fbproduct.php]: Why is this line needed?

File: includes/fbproduct.php [hunk 2]
```diff
@@ -35,6 +35,7 @@ class WC_Facebook_Product {
 	const FB_VARIANT_IMAGE       = 'fb_image';
 	const FB_VISIBILITY          = 'fb_visibility';
 	const FB_REMOVE_FROM_SYNC    = 'fb_remove_from_sync';
+	const FB_RICH_TEXT_DESCRIPTION = 'fb_rich_text_description';
```
Comment (mshymon) [inline includes/fbproduct.php:39]: Why do we need a variable for rich text description here and why we would want oto save it to the meta data? As we can populate it on the fly when needed from the woo data model. 
The "fb_product_description" is needed because we have a a text box in UI, so we have to save what users enters there. But in case of rich text description, we don't have it, unless we add UI of a rich text into existing "FB Description" text box.

File: includes/fbproduct.php [hunk 3]
```diff
@@ -447,6 +467,58 @@ public function get_fb_description() {
...
+		if ( empty( $rich_text_description ) ) {
+			// Try to get description from post meta if fb description has been set
+			$rich_text_description = get_post_meta(
+				$this->id,
+				self::FB_PRODUCT_DESCRIPTION,
+				true
+			);
+		}
+
+		// Check if the product type is a variation and no rich text description is found yet
```
Comment (mshymon) [inline includes/fbproduct.php]: Rather then just saying what the check are, better to say what behaviour are we trying to impose, e.g. "In case of a variable product we want to take rich text description of the variant, and when it is not available fallback on the main (parent) product rich text description"

**Review by mshymon — CHANGES_REQUESTED** [review]

File: includes/fbproduct.php [hunk 4]
```diff
@@ -35,13 +35,17 @@ class WC_Facebook_Product {
...
 
 	const MIN_DATE_1 = '1970-01-29';
 	const MIN_DATE_2 = '1970-01-30';
 	const MAX_DATE   = '2038-01-17';
 	const MAX_TIME   = 'T23:59+00:00';
 	const MIN_TIME   = 'T00:00+00:00';
 
+    public static $rich_text_description_source = WC_Facebookcommerce_Utils::WOO_DESCRIPTION;
+	
+
```
Comment (vink28) [inline includes/fbproduct.php]: Remove extra blank lines

File: includes/fbproduct.php [hunk 5]
```diff
@@ -652,8 +725,13 @@ public function prepare_product( $retailer_id = null, $type_to_prepare_for = sel
 			$brand = wp_strip_all_tags( WC_Facebookcommerce_Utils::get_store_name() );
 		}
 
+		$rich_text_description = $this->get_rich_text_description();
+		
+		print_r('source', self::$rich_text_description_source);
```
Comment (vink28) [inline includes/fbproduct.php]: Lets remove the print statement as it is of no use for us as we don't have visibility on locally printed statements and currently we don't have rights to sync/access these logs. Also excessive printing may increase local log file size which may take up most of the storage capacity of hosted servers.
I'm sure there are numerous places in our plugin where we are adding debug logs, but we should understand how can we build process to timely clean up these logs automatically to avoid data storage overload.

File: includes/fbproduct.php [hunk 6]
```diff
@@ -357,6 +367,13 @@ public function set_product_image( $image ) {
 		}
 	}
 
+	public function set_rich_text_description( $rich_text_description ) {
```
Comment (vink28) [inline includes/fbproduct.php:374]: What is use of this ? Unless we are planning to locally store rich_text_description corresponding to each product then we don't need to create a setter explicitly here.
Our logic of deriving rich_text_description is from existing html tagged product description and send it to our APIs for syncing its values rather explicitly persisting it locally. 
Also locally saving this is not a good idea as it is already beeing saved and we may not want to save the value again just with a new fb specific key.
So lets remove it!

File: includes/fbproduct.php [hunk 7]
```diff
@@ -447,6 +464,62 @@ public function get_fb_description() {
...
+
+		if ( $this->fb_description ) {
+			$rich_text_description = $this->fb_description;
+		}
+
+		if ( $this->rich_text_description ) {
+			$rich_text_description = $this->rich_text_description;
+		}
+		
+
```
Comment (vink28) [inline includes/fbproduct.php]: Remove extra lines

File: includes/fbproduct.php [hunk 8]
```diff
@@ -447,6 +464,62 @@ public function get_fb_description() {
 		return apply_filters( 'facebook_for_woocommerce_fb_product_description', $description, $this->id );
 	}
 
+	public function get_rich_text_description() {
+		$rich_text_description = '';
+
+		if ( $this->fb_description ) {
```
Comment (vink28) [inline includes/fbproduct.php]: **Code Commenting Best Practices**
When writing code, it's essential to include comments that explain the logic behind our decisions. This is especially important when determining the preferred product description, as advertisers often rely on code comments to understand the behavior of each product attribute.
Something like:
// Determine the preferred product description based on the following logic:
// 1. Check if the rich text description is available and not empty.
// 2. If the rich text description is available, use it as the preferred description.
// 3. Otherwise, fall back to the plain text description.

By adding comments like this, we can make our code more readable and maintainable, and provide better documentation for advertisers who rely on our code comments to understand the behavior of our product attributes.

File: includes/fbproduct.php [hunk 9]
```diff
@@ -447,6 +464,62 @@ public function get_fb_description() {
...
+			$rich_text_description = WC_Facebookcommerce_Utils::clean_string($this->woo_product->get_rich_text_description());
+
+			// If the variant's rich text description is still empty, use the main product's rich text description as a fallback.
+			if (empty($rich_text_description) && $this->main_description) {
+				$rich_text_description = $this->rich_text_description;
+			}
+		}
+
+		// If no description is found from meta or variation, get from post
+		if ( empty( $rich_text_description ) ) {
```
Comment (vink28) [inline includes/fbproduct.php:521]: I can see there are a lot of conditions here, ideally each of these conditions should be covered by tests and clearly highlighted in dogfooding document and should be manually tested by QA.
I don't see variants and post_data conditions in the unit tests below. Please update the tests.

File: includes/fbproduct.php [hunk 10]
```diff
@@ -665,14 +743,17 @@ public function prepare_product( $retailer_id = null, $type_to_prepare_for = sel
 				'price'                 => $this->get_fb_price( true ),
 				'availability'          => $this->is_in_stock() ? 'in stock' : 'out of stock',
 				'visibility'            => Products::is_product_visible( $this->woo_product ) ? \WC_Facebookcommerce_Integration::FB_SHOP_PRODUCT_VISIBLE : \WC_Facebookcommerce_Integration::FB_SHOP_PRODUCT_HIDDEN,
```
Comment (vink28) [inline includes/fbproduct.php]: Fix the indentations. I know formatting is currently a pain, if you have bandwidth then please fix the code sniffing for this package by fixing the phpcs files which will automatically fix the code formatting as soon as you save the code changes.

File: includes/fbproduct.php [hunk 11]
```diff
@@ -696,7 +777,9 @@ public function prepare_product( $retailer_id = null, $type_to_prepare_for = sel
 				'currency'              => get_woocommerce_currency(),
 				'availability'          => $this->is_in_stock() ? 'in stock' : 'out of stock',
 				'visibility'            => Products::is_product_visible( $this->woo_product ) ? \WC_Facebookcommerce_Integration::FB_SHOP_PRODUCT_VISIBLE : \WC_Facebookcommerce_Integration::FB_SHOP_PRODUCT_HIDDEN,
-			);
+			),
+			self::$rich_text_description_source === WC_Facebookcommerce_Utils::WOO_DESCRIPTION ? array('rich_text_description' => $rich_text_description) : array()
```
Comment (vink28) [inline includes/fbproduct.php]: Formatting

File: includes/fbutils.php [hunk 12]
```diff
@@ -34,6 +34,9 @@ class WC_Facebookcommerce_Utils {
 		const FB_VARIANT_PATTERN = 'pattern';
 		const FB_VARIANT_GENDER  = 'gender';
 
+		const FB_DESCRIPTION = 'fb_description';
+		const WOO_DESCRIPTION = 'woo_description';
```
Comment (vink28) [inline includes/fbutils.php]: Lets avoid prefix like 'Woo' in our plugin codebase. Either use the WooCommerce's 'description' or use prefix like 'wc' only if it represents WooCommerce's value for description attribute

File: includes/fbutils.php [hunk 13]
```diff
@@ -269,6 +272,10 @@ public static function clean_string( $string ) {
 			 * @param bool   $apply_shortcodes Shortcodes are applied if set to `true` and stripped out if set to `false`.
 			 * @param string $string           String to clean up.
 			 */
+
+			 if (!$string){
```
Comment (vink28) [inline includes/fbutils.php]: Check if string is empty or null

**Review by vinkmeta — CHANGES_REQUESTED** [review]
Added multiple comments, please fix them!

File: includes/Admin/Enhanced_Catalog_Attribute_Fields.php [hunk 14]
```diff
@@ -47,7 +47,7 @@ class Enhanced_Catalog_Attribute_Fields {
 	 */
 	private $category_handler;
 
-	public function __construct( $page_type, \WP_Term $term = null, \WC_Product $product = null ) {
+	public function __construct( $page_type, ?\WP_Term $term = null, ?\WC_Product $product = null ) {
```
Comment (devbodaghe) [inline includes/Admin/Enhanced_Catalog_Attribute_Fields.php:50]: Implicitly marking parameter $term as nullable is deprecated, the explicit nullable type must be used instead

This change ensures that the parameters $term and $product can explicitly accept null values.

File: includes/Admin.php [hunk 15]
```diff
@@ -1355,20 +1364,25 @@ public function add_product_variation_edit_fields( $index, $variation_data, $pos
 			)
 		);
 
-		woocommerce_wp_textarea_input(
+		echo '<div class="wp-editor-wrap">';
+		echo '<label for="' . esc_attr(\WC_Facebookcommerce_Integration::FB_PRODUCT_DESCRIPTION) . '">' . 
+			 esc_html__( 'Facebook Description', 'facebook-for-woocommerce' ) . 
+			 '</label>';
+		wp_editor(
+			$rich_text_description,
```
Comment (devbodaghe) [inline includes/Admin.php:1371]: this is the major reason for preserving rich text description as on the ui, we need to render the text in the box with styling added

File: includes/fbproduct.php [hunk 16]
```diff
@@ -448,15 +472,81 @@ public function get_fb_description() {
...
+
+		// Check if the fb description is set as that takes preference
+		if ( $this->fb_description ) {
+			$rich_text_description = $this->fb_description;
+		}
+
+		// If the rich text description is available, use it as the preferred description
+		if ( $this->rich_text_description ) {
+			$rich_text_description = $this->rich_text_description;
+		}
```
Comment (mshymon) [inline includes/fbproduct.php]: nit. Why do we set one value above and then override it immediately? It is easier to follow straightforward logic if ... else if  ...  else if... and so on to set the right value.

File: includes/fbproduct.php [hunk 17]
```diff
@@ -448,15 +472,81 @@ public function get_fb_description() {
...
+
+		// Try to get description from post meta if fb description has been set
```
Comment (mshymon) [inline includes/fbproduct.php]: has not been set?

File: tests/Unit/fbproductTest.php [hunk 18]
```diff
@@ -360,4 +360,216 @@ public function test_gtin_for_variable_product_unset() {
...
+		$this->assertEquals( $rich_text_description,  'rich text description' );
+	}	
+	
+	/**
+	 * Test html tags preservation for rich text description
+	 * @return void
+	 */
+	public function test_html_preservation_for_rich_text_description() {
+    // Create a simple product
+    $product = WC_Helper_Product::create_simple_product();
```
Comment (mshymon) [inline tests/Unit/fbproductTest.php]: The alignment of raws looks off

File: tests/Unit/fbproductTest.php [hunk 19]
```diff
@@ -360,4 +360,216 @@ public function test_gtin_for_variable_product_unset() {
...
+    // Assert that the HTML content is preserved
+    $this->assertEquals($htmlContent, $richTextDescription);
+	}
+
+	/**
+	 * Tests for get_rich_text_description() method
+	 */
+	public function test_get_rich_text_description() {
+		// Test 1: Gets rich text description from fb_description if set
+		$product = WC_Helper_Product::create_simple_product();
```
Comment (mshymon) [inline tests/Unit/fbproductTest.php:384]: nit. $woo_product for easy comprehension of reading

File: tests/Unit/fbproductTest.php [hunk 20]
```diff
@@ -360,4 +360,216 @@ public function test_gtin_for_variable_product_unset() {
...
+		// Test 5: Falls back to post content if no other description is set
+		$product = WC_Helper_Product::create_simple_product();
+		$product->set_description('<p>product content description</p>');
+		$product->save();
+		
+		$facebook_product = new \WC_Facebook_Product($product);
+		$description = $facebook_product->get_rich_text_description();
+		$this->assertEquals('<p>product content description</p>', $description);
+
+		// Test 6: Falls back to post excerpt if content is empty and sync_short_description is true
```
Comment (mshymon) [inline tests/Unit/fbproductTest.php:422]: nit. excerpt --> except

File: facebook-commerce.php [hunk 21]
```diff
@@ -135,7 +135,7 @@ class WC_Facebookcommerce_Integration extends WC_Integration {
 	public const FB_PRODUCT_GROUP_ID    = 'fb_product_group_id';
 	public const FB_PRODUCT_ITEM_ID     = 'fb_product_item_id';
 	public const FB_PRODUCT_DESCRIPTION = 'fb_product_description';
-
+	public const FB_RICH_TEXT_DESCRIPTION = 'fb_rich_text_description';
```
Comment (mshymon) [inline facebook-commerce.php:138]: Do we need both FB_PRODUCT_DESCRIPTION  and FB_PRODUCT_DESCRIPTION ?

File: includes/fbproduct.php (hunk 6 above)
Comment (mshymon) [reply includes/fbproduct.php:374]: @devbodaghe please reply to this comment

File: tests/Unit/fbproductTest.php [hunk 22]
```diff
@@ -360,4 +360,216 @@ public function test_gtin_for_variable_product_unset() {
 
 		$this->assertEquals(isset($data['gtin']), false);
 	}
+
+		
+	/**
+	 * Test it gets rich text description from post meta.
+	 * @return void
+	 */
+	public function test_get_rich_text_description_from_post_meta() {
```
Comment (mshymon) [inline tests/Unit/fbproductTest.php:369]: "_from_post_meta" rather then saying what is the storage of the data, I think better operate at level of data input at user level, e.g. "woo product description", "woo short product description", "FB product description"

**Review by mshymon — COMMENTED** [review]
Thanks for making this work, @devbodaghe!

I left some comments to consider.

The getting rich description function looks complex and it is difficult to follow, try to simplify if possible.

Also, let's ask QA to do testing final version before we merge this.

File: tests/Unit/fbproductTest.php (hunk 19 above)
Comment (devbodaghe) [reply tests/Unit/fbproductTest.php:384]: https://github.com/devbodaghe/facebook-for-woocommerce/pull/1#discussion_r1857474267

Mentions to not use woo in codebase
Will update to `wc_product`

File: facebook-commerce.php (hunk 21 above)
Comment (devbodaghe) [reply facebook-commerce.php:138]: I imagine you mean `FB_RICH_TEXT_DESCRIPTION` and `FB_PRODUCT_DESCRIPTION`

Yes. One holds the rich value of the input and the other holds the plain value
